Migrate Parameter Parsing: introducing `ParameterAccessor` (#341)
enzo-project/enzo-e: doc/source/devel/param.rst

Essentially, parameters are organized within groups and (possibly subgroups).
In other words, the parameters can be thought of as leaf nodes in a tree-hierarchy of "groups".
The following snippet illustrates the organization of parameters from a hypothetical parameter file:

 ::

  Method {
     list = [ "mhd_vlct", "grackle"];

     mhd_vlct {
        # other assorted parameters...
        courant = 0.3;
     }

     grackle {
        # other assorted parameters...
        courant = 0.4;
     }
  }

The organization of parameters in a group hierarchy is analogous to the organization of files in a directory hierarchy.
Continuing this analogy, we have devised shorthand for naming parameters in the documentation (and throughout the codebase) that is similar to a file path.
One can think of these names as a "parameter-path".

- documentation refering to the parameters in the above snippet would mention :par:paramfmt:`Method:list`, :par:paramfmt:`Method:mhd_vlct:courant`, or :par:paramfmt:`Method:grackle:courant`.

> - in the codebase, one could see strings refering to ``"Method:list"``, ``"Method:mhd_vlct:courant"``, ``"Method:grackle:courant"`` (in practice, you may not see these particular strings).

Author: good point

   Alternative logic may be needed to the above expressions in slightly more sophisticated cases (for example if the parameter expects a list of values or if you want to abort the program if the parameter can't be found).

.. note::

    In some cases, parameter-values may be retrieved from the parameter-file in a static factory method, typically called ``from_parameters``, rather than in a constructor.

    There are sometimes cases where the use of factory-method could be very useful (that should usually justified in a comment).
    But factory-methods are most commonly used in cases where it simplified the transition from the older parameter-parsing approach to this newer approach.

==============================
Design Overview (new approach)
==============================

Our new approach revolves around the usage of the :cpp:class:`!ParameterAccessor` class for accessing/querying parameters stored in an instance of the :cpp:class:`Parameters` class.
Instances of the :cpp:class:`!ParameterAccessor` class are light-weight and are expected to have a short-lifetime (akin to :cpp:class:`!Field` or :cpp:class:`!Particle`).

> **As illustrated above, instances of the** :cpp:class:`!ParameterAccessor` **class are expected to be passed to the constructor (or factory-method) of classes that inherit from Cello class-hierarchy.**

Author: Good point! It's not illustrated at all. I added an example

   - ``p.value_logical(s, false)`` if the parameter is expected to specify a boolean value and if it defaults to a value of ``false`` when the parameter is not specified.

   - ``p.value_integer(s, 7)`` if the parameter is expected to be an integer and if it defaults to a value of ``7`` when the parameter is not specified.

   - ``p.value_float(s, 2.0)`` if the parameter is expected to be a floating-point value and if it defaults to a value of ``2.0`` when the parameter is not specified.

   - ``p.value_string(s, "N/A")`` if the parameter is expected to be a string and if it defaults to a value of ``"N/A"`` when the parameter is not specified.

If the user specified the desired parameter, but with an unexpected type, the program will abort with an error message (another function is provided to query the parameter type)

In each of these snippets, ``s`` is **always** a parameter path, but the precise interpretation depends on how ``p`` is defined.
When ``p`` references a :cpp:class:`!Parameters` instance, ``s`` must specify an absolute parameter-path.

> In contrast, when ``p`` references a :cpp:class:`!ParameterGroup` instance, ``s`` must specify the path relative to a 

Reviewer: This sentence is missing something.

For completeness, consider the following parameter-file snippet:

 ::

  Physics {
    fluid_props {
      eos    {  gamma = 1.6666666666666667; }
      floors {  density  = 1.0e-10;         }
    }
  }

> We will now highlight what the choices of ``s`` that specify is :par:paramfmt:`Physics:fluid_props:eos:gamma` for different choices of ``p``.

Reviewer: ```suggestion
The table below clarifies the value of ``s`` that should be used to specify :par:paramfmt:`Physics:fluid_props:eos:gamma` for different choices of ``p``.
```

| :par:paramfmt:`Physics:fluid_props:floors` |                                     |
+--------------------------------------------+-------------------------------------+

Common Patterns in the Codebase
-------------------------------

Enzo-E and Cello define many classes descended from the Cello-class-hierarchy (e.g. :cpp:class:`!Method` subclasses) that are directly initialized from the parameters in a single parameter-group.
These classes are commonly initialized with a constructor that accepts a :cpp:class:`!ParametersGroup` instance (associated with the appropriate parameter-group) as an argument.

.. COMMENT-BLOCK

    In the future, we may want to swap out the example-case to something other that a
    Method class (once we start using directly passing ParameterGroup to other types)

For the sake of example, let's consider the :cpp:class:`!EnzoMethodHeat` class.
This class is configured by parameters like the ones in the :par:paramfmt:`Method:heat` group from the following parameter-file snippet:

 ::

  Method {
     list = [ "heat"];

     heat {
        alpha = 0.6;
        courant = 0.3;
     }
  }

> Here's we present (a heavily editted) example of what the class's constructor might look like:

Reviewer: ```suggestion
Here's we present (an edited) example of what the class's constructor might look like:
```

In rare cases (e.g. during refactoring when we convert a previously Method-specific parameter to a Physics parameter and want to retain backwards compatability), exceptions to this philosophy need to be made.
Thus, an "escape-hatch" is provided to directly access the global :cpp:class:`!Parameter` object: call the :cpp:expr:`cello::parameters()`.
Please, avoid using this "escape-hatch" unless it's truly necessary.

.. todo::

   We could consider extending the analogy between a parameter-path and a file path.
   For example, one could imagine interpreting a path that begins with a ``:`` as an absolute parameter-path and all other strings as relative parameter-paths.

   This would probably streamline the documentation to some degree.

   If we were to do that, we would need to modify the code to recognize this convention.
   We would probably also want to modify the various parameter-accessor methods of the :cpp:class:`!ParameterGroup` to continue to restrict access to parameters within the common root-path that a :cpp:class:`!ParameterGroup` is configured with.

=================

> Historic Approach

Reviewer: ```suggestion
Historical Approach
```
